File: solutions/python3/HeightChecker.py
def counting_sort(arr: List[int]):
    freqs = defaultdict(int)
    max_val = min_val = arr[0]
    for num in arr:
        freqs[num] += 1
        if max_val < num:
            max_val = num
        if num < min_val:
            min_val = num
    idx = 0
    for val in range(min_val, max_val + 1):
        while freqs.get(val, 0) > 0:
            arr[idx] = val
            freqs[val] -= 1
            idx += 1


class Solution:
    def heightChecker(self, heights: List[int]) -> int:
        temp = deepcopy(heights)
        # temp.sort()
        counting_sort(temp)
        res = 0
        for i in range(len(heights)):
            if heights[i] != temp[i]:
                res += 1
        return res
        

File: solutions/java/HeightChecker.java
class Solution {
    public int heightChecker(int[] heights) {
        int res = 0;
        int[] clone = heights.clone();
        countingSort(clone);
        for (int i = 0; i < heights.length; i++){
            if (heights[i] != clone[i])
                res++;
        }
        return res;
    }

    private void countingSort(int[] arr){
        int maxVal, minVal;
        maxVal = minVal = arr[0];
        HashMap<Integer, Integer> freqs = new HashMap<>();
        for (int num : arr){
            freqs.put(num, freqs.getOrDefault(num, 0) + 1);
            maxVal = Math.max(maxVal, num);
            minVal = Math.min(minVal, num);
        }
        int idx = 0;
        for (int val = minVal; val <= maxVal; val++){
            while (freqs.getOrDefault(val, 0) > 0){
                arr[idx] = val;
                freqs.put(val, freqs.get(val) - 1);
                idx++;
            }
        }
    }
}